### PR TITLE
♻️🐛💄 Overhaul tests

### LINKED ISSUE #53: Missing dependency in package.json
the `utils.ts` uses bluebird, but bluebird isn't listed in addict-iocs dependencies

### LINKED ISSUE #58: Throw a meaningful error, if undefined is registered
I just had a typo when registering a class to the ioc container. Because of this, i unknowingly registered `undefined` as a class to the ioc container.

This did not throw an error when reigstering, but it did throw an error when it was tried to be resolved:

```
TypeError: type is not a constructor
    at ConfigResolver.Resolver._createInstanceByConstructor (/Users/heiko/Desktop/Projekte/BPMN/consumer_api_meta/node_modules/addict-ioc/dist/commonjs/resolver.js:151:24)
    at ConfigResolver.Resolver._createInstance (/Users/heiko/Desktop/Projekte/BPMN/consumer_api_meta/node_modules/addict-ioc/dist/commonjs/resolver.js:136:29)
    at ConfigResolver.Resolver.createInstance (/Users/heiko/Desktop/Projekte/BPMN/consumer_api_meta/node_modules/addict-ioc/dist/commonjs/resolver.js:129:21)
    at InvocationContainer.Container._createType (/Users/heiko/Desktop/Projekte/BPMN/consumer_api_meta/node_modules/addict-ioc/dist/commonjs/container.js:531:32)
    at InvocationContainer.<anonymous> (/Users/heiko/Desktop/Projekte/BPMN/consumer_api_meta/node_modules/addict-ioc/dist/commonjs/container.js:387:41)
    at step (/Users/heiko/Desktop/Projekte/BPMN/consumer_api_meta/node_modules/addict-ioc/dist/commonjs/container.js:50:23)
    at Object.next (/Users/heiko/Desktop/Projekte/BPMN/consumer_api_meta/node_modules/addict-ioc/dist/commonjs/container.js:31:53)
    at fulfilled (/Users/heiko/Desktop/Projekte/BPMN/consumer_api_meta/node_modules/addict-ioc/dist/commonjs/container.js:22:58)
```

It would be nice if addict-ioc would print a warning, when `undefined` or `null` is registered to the container. This warning should include the registration key

### LINKED ISSUE #61: 🐛 getKeysByTags returns duplicated entries, when searching for multiple tags
#### Describe your issue

Let's say you have the following scenario:

```js
container
  .register('keyA' TypeA)
  .tags('tag1', 'tag2')

container
  .register('keyB' TypeB)
  .tags('tag1')
```

When calling `container.getKeysByTags('tag1', 'tag2')`, I'd expect the following result:
```js
['keyA']
```

However, what you get is this:

```js
['keyA', 'keyA', 'keyB']
```

I assume this is because the `getKeysByTags` function doesn't check if the result set already contains the key of a matching registration.

#### Possible Solution

Add a filter to the `getKeysByTags` method, which checks if a key is already contained in the result set.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### LINKED ISSUE #62: 🐛 Object.assign() causes tests to use faulty data, due to the object references it creates
#### Describe your issue

While running the tests for overwriting keys, I discovered, that some later tests will contain the settings for overwritten keys that previous tests have generated.

This is the same effect we had trouble with in the ProcessEngine and is most likely due to the usage of `Object.assign`, which creates hard references to the object to which properties are being assigned.

#### Reproduction steps

Easiest is to run the following tests that use overwritten keys in sequence:
- `container/validate_dependencies_test`
- `registration/overwrite_test`

When the assertions for the second tests are being performed, you will get errors which distinctly show the `overwrittenKeys` settings from the first tests.

#### Possible solution

In the ProcessEngines packages, we had great success with using `clone` intead of `Object.assign`.
We should do the same here.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by ElRaptorus**
TL DR:
Make all tests great again.

**Changes:**

Tests:
1. Update all tests so they match the current implementation of the addict-ioc package.
2. Remove tests that cover functions which no longer exist.

Fixes:
1. Fix the "getKeysByTags" function: When retrieving registrations for multiple keys, a registration that has more than one matching key was returned multiple times (See #61).
2. Use `clone` for creating copies of an object, instead of `Object.assign` to prevent the original object references from being overwritten (See #62).

Enhancements:
1. Replace obsolete `node-uuid` package with `uuid`.
2. Add checks to `register` methods, which ensure that `key` is a string and `type` is not null or undefined (See  #58).

Miscellaneous:
1. Fix styling for all tests and .ts components.
2. Update the ESLint rules to match the ones we use in the ProcessEngine (we should consider updating the 5minds ESLint style).
3. Enhance Contributors info.
4. Use fixed dependency versions.
5. Remove `bluebird`, since its no longer used anywhere.

**Issues:**

Closes #53 
Closes #58
Closes #61 
Closes #62 

PR: #60

## How can others test the changes?

Run the tests.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: test/container/get_keys_by_tags_test.js [hunk 1]
```diff
@@ -106,155 +103,74 @@ describe('Dependency Injection Container Get Keys By Tags Test', () => {
...
-    const attributeQuery = {
-      [testTag]: testTagValue,
-      [testTagTwo]: testTagTwoValue
-    };
-
-    const foundKeys = container.getKeysByAttributes(attributeQuery);
+    const foundKeys = container.getKeysByTags(attributeQuery);
 
-    should(foundKeys.length).equal(0);
+    should(foundKeys.length).be.equal(1);
```
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/container/get_keys_by_tags_test.js]: Fixed in 9041b59

File: test/registration/bind_functions_test.js [hunk 2]
```diff
@@ -6,31 +6,43 @@ const Container = require('./../../dist/commonjs').Container;
...
+    const registration = container.getRegistration(testTypeKey);
+    const bindFunctionsSetting = registration.settings.functionsToBind;
+
+    should(bindFunctionsSetting).be.an.Array();
+    should(bindFunctionsSetting.length).be.equal(0);
   });
 
-  it('should throw error if no registration is started', function testCallback(next) {
+  it('should successfully register the given type with the given function stored in functionsToBind', () => {
+
```
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/registration/bind_functions_test.js]: This feature doesn't seem to be implemented right now anyway.
The function names stored in the `functionsToBind` setting are never used anywhere.

I can adjust the tests here and then place the entire thing on `skip` until the implementation is finished.

File: test/container/resolve/errors_test.js [hunk 3]
```diff
@@ -6,122 +6,128 @@ const Container = require('./../../../dist/commonjs').Container;
...
+      const result = container.resolve(key);
+      should.fail(result, 'error', 'This should have failed, because of a circular dependency!');
     } catch (error) {
-      should(error).not.be.null();
-      next();
+      should.exist(error);
+      // TODO: Right now, this causes a "Maximum Call Stack Size exceeded" error.
+      // When a proper error is thrown, we can assert the error message here.
+      //
+      // Note that a proper error is thrown, when "validateDependencies" is called.
```
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/container/resolve/errors_test.js]: The error doesn't occur when calling `validateDependencis`, but when calling `resolve()`. 
Maybe I'm missing something here, but there doesn't seem to be a validation in place for that method.

File: test/container/resolve/errors_test.js [hunk 4]
```diff
@@ -6,122 +6,128 @@ const Container = require('./../../../dist/commonjs').Container;
...
-      container.resolve(key);
-
+      const result = container.resolve(key);
+      should.fail(result, 'error', 'This should have failed, because of a circular dependency!');
     } catch (error) {
-      should(error).not.be.null();
-      next();
+      should.exist(error);
+      // TODO: Right now, this causes a "Maximum Call Stack Size exceeded" error.
+      // When a proper error is thrown, we can assert the error message here.
```
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/container/resolve/errors_test.js]: dito.

File: test/container/resolve/errors_test.js [hunk 5]
```diff
@@ -170,108 +178,118 @@ describe('Dependency Injection Container Resolve Error Test', function describeC
...
+      .configure(() => undefined); //eslint-disable-line
 
+    try {
+      const result = container.resolve(key);
+      should.fail(result, 'error', 'This should have failed, because the config function returns undefined!');
     } catch (error) {
-      should(error).not.be.null();
-      next();
+      should.exist(error);
+      // TODO: Not sure if this case requires a dedicated message?
```
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/container/resolve/errors_test.js]: Actually, I dug around a little more and came to the conclusion that there is not much point in doing this test, because it would require is to provide a custom resolver.
The default resolver only returns the config as it is, which always results in success, as long as the config-property on the target instance exists.

I don't think its a good idea to test customized behavior here. That should be done by whatever application uses a customized resolver.

File: test/registration/bind_functions_test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/registration/bind_functions_test.js]: Updated in:

https://github.com/5minds/addict-ioc/commit/39aacee0bd83e4057d22c5f3c921586bd56b995c#diff-09a127c61c67822308a0aa82851d7277R38

File: test/container/validate_dependencies_test.js [hunk 6]
```diff
@@ -6,263 +6,207 @@ const Container = require('./../../dist/commonjs').Container;
...
-      container.validateDependencies();
+    container.register(secondTestTypeKey, SecondaryTestType);
 
-    } catch (error) {
-      should(error).not.be.null();
-      next();
-    }
+    container
+      .register(primaryTestTypeKey, PrimaryTestType)
+      .overwrite(primaryTestTypeKey, secondTestTypeKey);
```
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/container/validate_dependencies_test.js]: Actually there is only a call to `validateDependencies()` missing. :3

[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/container/validate_dependencies_test.js]: Fixed in https://github.com/5minds/addict-ioc/pull/60/commits/3c6d99d81be51c5f2758abbcb2b47b8510d72f5b

File: test/container/set_config_provider_test.js [hunk 7]
```diff
@@ -6,42 +6,57 @@ const Container = require('./../../dist/commonjs').Container;
...
     const testConfig = {
-      test: '1'
+      test: '1',
     };
     const testConfigProvider = () => {
       return testConfig;
     };
 
-    container.setConfigProvider(testConfigProvider);
+    container.settings.resolver = testConfigProvider;
```
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/container/set_config_provider_test.js]: I'm not so sure that there is a need to do this kind of test at all, because using a customized resolver isn't exactly internal to the addict-ioc package.

Do we really want to test application-specific stuff here? 

See my other comment on this here: 
https://github.com/5minds/addict-ioc/pull/60#discussion_r228860389

File: test/container/set_config_provider_test.js [hunk 8]
```diff
@@ -6,42 +6,57 @@ const Container = require('./../../dist/commonjs').Container;
...
+  it('should throw an error, when an invalid config provider is assigned to the container', () => {
 
+    const testTypeKey = 'key';
     const invalidConfigProvider = {
-      something: 'false'
+      something: 'false',
     };
 
     try {
+      container.settings.resolver = invalidConfigProvider;
```
[the remark this answers is not in the document]
Comment (ElRaptorus) [reply test/container/set_config_provider_test.js]: see above.

File: test/container/resolve/errors_test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (5ebastianMeier) [reply test/container/resolve/errors_test.js]: I'll have to look into this further. Can't find the problem right away and this shouldn't block the PR as a whole.

File: test/container/resolve/errors_test.js [hunk 9]
```diff
@@ -170,108 +144,95 @@ describe('Dependency Injection Container Resolve Error Test', function describeC
...
         this._config = value;
       }
-    };
 
-    container.register(dependencyKey, TestType);
+    }
+
+    // This must be done here, because the resolver uses the instances prototype to validate the injection target.
+    // But the field won't be set until after the constructor was called, so we wouldn't get the error message we desire.
+    SecondType.prototype.injectionTargetMethod = [];
```
Comment (5ebastianMeier) [inline test/container/resolve/errors_test.js]: In other words it doesn't find the field if its a property or a method declared on class level? If so this is not the intended behavior.

File: test/registration/configure_test.js [hunk 10]
```diff
@@ -7,46 +7,62 @@ const Container = require('./../../dist/commonjs').Container;
...
 
+    should(configSetting).be.eql(config);
   });
 
-  it('should throw error if type of config declaration is not supported', function testCallback(next) {
+  it('should successfully create a configured registration, even when an ususal config is passed', () => {
+
+    const unsupportedConfig = 123456;
+
+    // Note that this should succeed, because validation is performed during the "validateDependencies" check.
```
Comment (5ebastianMeier) [inline test/registration/configure_test.js]: 👍

File: test/registration/overwrite_test.js [hunk 11]
```diff
@@ -6,58 +6,55 @@ const Container = require('./../../dist/commonjs').Container;
...
+    should(overwriteSetting).be.eql(expectedSetting);
   });
 
-  it('should throw error if no dependency for overwritten key is declared ', function testCallback(next) {
+  it('should not throw an error, if no dependency for the overwritten key is declared ', () => {
 
-    try {
-      const key = 'test';
-      const secondKey = 'secondKey';
+    // Note that validating this is handled during "validateDependencies".
```
Comment (5ebastianMeier) [inline test/registration/overwrite_test.js]: 👍
